src/Microsoft.DotNet.XHarness.iOS/AppRunner.cs

```diff
@@ -0,0 +1,359 @@
+using Microsoft.DotNet.XHarness.iOS.Shared;
+using Microsoft.DotNet.XHarness.iOS.Shared.Execution;
+using Microsoft.DotNet.XHarness.iOS.Shared.Execution.Mlaunch;
+using Microsoft.DotNet.XHarness.iOS.Shared.Hardware;
+using Microsoft.DotNet.XHarness.iOS.Shared.Listeners;
+using Microsoft.DotNet.XHarness.iOS.Shared.Logging;
+using Microsoft.DotNet.XHarness.iOS.Shared.Utilities;
+
+namespace Microsoft.DotNet.XHarness.iOS
+{
+    public class AppRunner
+    {
+        private readonly IProcessManager _processManager;
+        private readonly IHardwareDeviceLoader _hardwareDeviceLoader;
+        private readonly ISimulatorLoader _simulatorLoader;
+        private readonly ISimpleListenerFactory _listenerFactory;
+        private readonly ICrashSnapshotReporterFactory _snapshotReporterFactory;
+        private readonly ICaptureLogFactory _captureLogFactory;
+        private readonly IDeviceLogCapturerFactory _deviceLogCapturerFactory;
+        private readonly ITestReporterFactory _testReporterFactory;
+        private readonly ILogs _logs;
+        private readonly ILog _mainLog;
+        private readonly IHelpers _helpers;
+
+        public AppRunner(IProcessManager processManager,
+            IHardwareDeviceLoader hardwareDeviceLoader,
+            ISimulatorLoader simulatorLoader,
+            ISimpleListenerFactory simpleListenerFactory,
+            ICrashSnapshotReporterFactory snapshotReporterFactory,
+            ICaptureLogFactory captureLogFactory,
+            IDeviceLogCapturerFactory deviceLogCapturerFactory,
+            ITestReporterFactory reporterFactory,
+            ILog mainLog,
+            ILogs logs,
+            IHelpers helpers)
+        {
+            _processManager = processManager ?? throw new ArgumentNullException(nameof(processManager));
+            _hardwareDeviceLoader = hardwareDeviceLoader ?? throw new ArgumentNullException(nameof(hardwareDeviceLoader));
+            _simulatorLoader = simulatorLoader ?? throw new ArgumentNullException(nameof(simulatorLoader));
+            _listenerFactory = simpleListenerFactory ?? throw new ArgumentNullException(nameof(simpleListenerFactory));
+            _snapshotReporterFactory = snapshotReporterFactory ?? throw new ArgumentNullException(nameof(snapshotReporterFactory));
+            _captureLogFactory = captureLogFactory ?? throw new ArgumentNullException(nameof(captureLogFactory));
+            _deviceLogCapturerFactory = deviceLogCapturerFactory ?? throw new ArgumentNullException(nameof(deviceLogCapturerFactory));
+            _testReporterFactory = reporterFactory ?? throw new ArgumentNullException(nameof(_testReporterFactory));
+            _mainLog = mainLog ?? throw new ArgumentNullException(nameof(mainLog));
+            _logs = logs ?? throw new ArgumentNullException(nameof(logs));
+            _helpers = helpers ?? throw new ArgumentNullException(nameof(helpers));
+        }
+
+        public async Task<(string deviceName, int exitCode)> RunApp(
```

Comment:
Suggestion:  This one function is 302 lines long.  Seems like some of it could be factored into helper functions for readability (and perhaps in the future, test-ability)

Reply:
I totally agree. This code is not my invention unfortunately and was migrated from Xamarin:
https://github.com/xamarin/xamarin-macios/blob/master/tests/xharness/AppRunner.cs

I didn't change it in this iteration because I have unit tests that I have it running against and was making sure I didn't break stuff.

I will definitely split this later into device/simulator part, if there's time..

Reply:
Makes sense, thanks for explaining.